keras/backend/common/dtypes.py

```diff
@@ -0,0 +1,310 @@
+}
+BIT64_TO_BIT32_DTYPE = {
+    "int64": "int32",
+    "uint64": "uint32",
+    "float64": "float32",
+    # TODO: support complex dtypes
+    # "complex128": "complex64",
+}
+
+
+@functools.lru_cache(maxsize=None)
+def canonicalize_dtype(float_type, dtype):
+    if float_type == "float16":
+        return BIT64_TO_BIT16_DTYPE.get(dtype, dtype)
+    elif float_type == "float32":
+        return BIT64_TO_BIT32_DTYPE.get(dtype, dtype)
+    elif float_type == "float64":
+        return dtype
+    else:
+        raise ValueError(
+            f"Invalid float_type: {float_type}, `float_type` must be one of"
+            f"('float16', 'float32', 'float64')."
+        )
+
+
+def dtype(x, *, canonicalize=False):
+    """Return the dtype object for a tensor or type."""
+    dtype = None
+    if x is None:
+        raise ValueError(f"Invalid argument to dtype: {x}.")
+    # when inputing python scalar or tensor
+    elif type(x) in PYTHON_SCALAR_DTYPES:
+        dtype = PYTHON_SCALAR_DTYPES[type(x)]
+    elif backend.is_tensor(x):
+        dtype = backend.standardize_dtype(x.dtype)
+    # when inputing python dtype or tensor dtype
+    elif isinstance(x, type) and x in PYTHON_SCALAR_DTYPES:
+        dtype = PYTHON_SCALAR_DTYPES[x]
+    elif hasattr(x, "name"):
+        dtype = x.name
+    elif hasattr(x, "__str__") and "torch" in str(dtype):
+        dtype = str(x).split(".")[-1]
+    # when inputing dtype string
+    elif x in ALLOWED_DTYPES:
+        dtype = x
+    elif x in ["int", "float"]:
+        if x == "int":
+            dtype = "int64"
+        else:
+            dtype = "float64"
+
+    if dtype is None or dtype not in ALLOWED_DTYPES:
+        raise ValueError(f"Invalid dtype: {x}")
+    return (
+        canonicalize_dtype(backend.floatx(), dtype) if canonicalize else dtype
+    )
+
+
+_bool_types = ["bool"]
+_int_types = [
+    "uint8",
+    "uint16",
+    "uint32",
+    "uint64",
+    "int8",
+    "int16",
+    "int32",
+    "int64",
+]
+_float_types = ["bfloat16", "float16", "float32", "float64"]
+_complex_types = ["complex64", "complex128"]
+_weak_types = ["int", "float", "complex"]
+_weak_types_python = [int, float, complex]
+
+
+def _default_types(dtype):
+    if not isinstance(dtype, str):
+        raise TypeError(f"Invalid type: {type(dtype)}. Must be a str.")
+    if dtype == "bfloat16":  # special case for bfloat16
+        dtype_indicator = "f"
+    else:
+        dtype_indicator = dtype[:1]
+
+    # get backend.floatx precision (16, 32, 64)
+    precision = backend.floatx()[-2:]
+
+    if dtype_indicator == "b":
+        return "bool"
+    elif dtype_indicator == "i":
+        return "int" + precision
+    elif dtype_indicator == "u":
+        return "uint" + precision
+    elif dtype_indicator == "f":
+        return "float" + precision
+    # TODO: support complex types
+    else:
+        raise ValueError(f"Invalid dtype: {dtype} that is failed to be parsed")
+
+
+def _keras_core_type(dtype, weak_type):
```

Comment:
You can remove references to keras_core and just use "keras"

Reply:
Done